pkg/pipeline/providers/gitlab/actions.go

```diff
@@ -0,0 +1,342 @@
+	}
+	remote, err := remote.New(sourceCodeProviderConfig)
+	if err != nil {
+		return nil, err
+	}
+	repos, err := remote.Repos(credential)
+	if err != nil {
+		return nil, err
+	}
+
+	//remove old repos
+	repositories, err := g.SourceCodeRepositoryLister.List(namespace, labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+	for _, repo := range repositories {
+		if repo.Spec.SourceCodeCredentialName == credentialID {
+			if err := g.SourceCodeRepositories.DeleteNamespaced(namespace, repo.Name, &metav1.DeleteOptions{}); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	//store new repos
+	for _, repo := range repos {
+		if !repo.Spec.Permissions.Admin {
+			//store only admin repos
+			continue
+		}
+		repo.Spec.SourceCodeCredentialName = credentialID
+		repo.Spec.UserName = credential.Spec.UserName
+		repo.Spec.SourceCodeType = credential.Spec.SourceCodeType
+		repo.Name = uuid.NewV4().String()
+		repo.Namespace = namespace
+		repo.Spec.ProjectName = credential.Spec.ProjectName
+		if _, err := g.SourceCodeRepositories.Create(&repo); err != nil {
+			return nil, err
+		}
+	}
+
+	return repositories, nil
+}
+
+func (g *GlProvider) authAddAccount(userID string, code string, config *v3.GitlabPipelineConfig) (*v3.SourceCodeCredential, error) {
+
+	if userID == "" {
+		return nil, errors.New("unauth")
+	}
+
+	remote, err := remote.New(config)
+	if err != nil {
+		return nil, err
+	}
+	account, err := remote.Login(code)
+	if err != nil {
+		return nil, err
+	}
+	account.Name = strings.ToLower(fmt.Sprintf("%s-%s-%s", config.Namespace, model.GitlabType, account.Spec.LoginName))
+	account.Namespace = userID
+	account.Spec.UserName = userID
+	account.Spec.ProjectName = config.ProjectName
+	_, err = g.SourceCodeCredentials.Create(account)
+	if apierror.IsAlreadyExists(err) {
+		exist, err := g.SourceCodeCredentialLister.Get(userID, account.Name)
+		if err != nil {
+			return nil, err
+		}
+		account.ResourceVersion = exist.ResourceVersion
+		return g.SourceCodeCredentials.Update(account)
+	} else if err != nil {
+		return nil, err
+	}
+	return account, nil
+}
+
+func (g *GlProvider) disableAction(request *types.APIContext) error {
+
+	ns, _ := ref.Parse(request.ID)
+	o, err := g.SourceCodeProviderConfigs.ObjectClient().UnstructuredClient().GetNamespaced(ns, model.GitlabType, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	u, _ := o.(runtime.Unstructured)
+	config := u.UnstructuredContent()
+	if e, ok := config[client.SourceCodeProviderConfigFieldEnabled].(bool); ok && e {
+		config[client.SourceCodeProviderConfigFieldEnabled] = false
+		if _, err := g.SourceCodeProviderConfigs.ObjectClient().Update(model.GitlabType, o); err != nil {
+			return err
+		}
+		if t, ok := config["projectName"].(string); ok && t != "" {
+			return g.cleanup(config["projectName"].(string))
+		}
+	}
+
+	return nil
+}
+
+func (g *GlProvider) cleanup(projectID string) error {
+
+	pipelines, err := g.PipelineLister.List("", labels.Everything())
```

Comment:
there is a way to add an indexer to the list; so you can request the objects for a particular project. It would save you from iterating over. The example of indexer can be found here: https://github.com/rancher/rancher/blob/9ad8d635cbbc9e79738d045c5dbee41aa5478238/pkg/controllers/user/rbac/project_handler.go#L58. Also check if there are any other places in the pipeline code that could use the same approach